rfcs/GraphQLOverWebSocket.md

```diff
@@ -0,0 +1,192 @@
+
+Messages are represented through the JSON structure and are stringified before being sent over the network. They are bidirectional, meaning both the server and the client must conform to the specified message structure.
+
+**All** messages contain the `type` field outlining the action this message describes. Depending on the type, the message can contain two more _optional_ fields:
+
+- `id` used for uniquely identifying server responses and connecting them with the client's requests
+- `payload` holding the extra "payload" information to go with the specific message type
+
+The server can close the socket (kick the client off) at any time. The close event dispatched by the server is used to describe the fatal error to the client.
+
+The client closes the socket and the connection by dispatching a `1000: Normal Closure` close event to the server indicating a normal closure.
+
+## Message types
+
+### `ConnectionInit`
+
+Direction: **Client -> Server**
+
+Indicates that the client wants to establish a connection within the existing socket. This connection is **not** the actual WebSocket communication channel, but is rather a frame within it asking the server to allow future operation requests.
+
+The server must receive the connection initialisation message within the allowed waiting time specified in the `connectionInitWaitTimeout` parameter during the server setup. If the client does not request a connection within the allowed timeout, the server will close the socket with the event: `4408: Connection initialisation timeout`.
+
+If the server receives more than one `ConnectionInit` message at any given time, the server will close the socket with the event `4429: Too many initialisation requests`.
+
+```typescript
+interface ConnectionInitMessage {
+  type: 'connection_init';
+  payload?: Record<string, unknown>;
+}
+```
+
+### `ConnectionAck`
+
+Direction: **Server -> Client**
+
+Expected response to the `ConnectionInit` message from the client acknowledging a successful connection with the server.
+
+The server can use the optional `payload` field to transfer additional details about the connection.
+
+```typescript
+interface ConnectionAckMessage {
+  type: 'connection_ack';
+  payload?: Record<string, unknown>;
+}
+```
+
+The client is now **ready** to request subscription operations.
+
+### `Subscribe`
+
+Direction: **Client -> Server**
+
+Requests an operation specified in the message `payload`. This message provides a unique ID field to connect published messages to the operation requested by this message.
+
+If there is already an active subscriber for an operation matching the provided ID, regardless of the operation type, the server must close the socket immediately with the event `4409: Subscriber for <unique-operation-id> already exists`.
+
+```typescript
+interface SubscribeMessage {
+  id: '<unique-operation-id>';
+  type: 'subscribe';
+  payload: {
+    operationName?: string | null;
+    query: string;
```

Comment:
@enisdenjo I have a question regarding persisted queries. At the moment the protocol puts the persisted query id in the query field which now poses the problem to the server that the server has to guess if the query field needs to be parsed or if it is a query id. I think it would be better to mirror the request properties form the http query to put the persisted query id in the extensions dictionary. This way the server does not need to guess and can provide more precise error messages.

Reply:
Agreed. My initial thinking was exclusive: you either use persisted queries or you do dont. Besides, I also encountered a few scenarios with users that want extensibility, the `extensions` field is the way to go.

Let me improve this.

Reply:
Awesome, we will adopt the protocol (as experimental) with hotchocolate for the June release. Awesome work!
https://github.com/ChilliCream/hotchocolate

Reply:
Landed in https://github.com/graphql/graphql-over-http/pull/140/commits/478e46f00f5904e3f66c59d99ab475e76dc38879 and released in [`graphql-ws@v4.7.0`](https://github.com/enisdenjo/graphql-ws/releases/tag/v4.7.0).

Wohoo, awesome stuff, looking forward to the June release, thank you!